// azure-functions-typescript/src/http.js
"use strict";
(function (HttpStatusCodes) {
    HttpStatusCodes[HttpStatusCodes["Continue"] = 100] = "Continue";
    HttpStatusCodes[HttpStatusCodes["SwitchingProtocols"] = 101] = "SwitchingProtocols";
    HttpStatusCodes[HttpStatusCodes["Processing"] = 102] = "Processing";
    HttpStatusCodes[HttpStatusCodes["Checkpoint"] = 103] = "Checkpoint";
    HttpStatusCodes[HttpStatusCodes["OK"] = 200] = "OK";
    HttpStatusCodes[HttpStatusCodes["Created"] = 201] = "Created";
    HttpStatusCodes[HttpStatusCodes["Accepted"] = 202] = "Accepted";
    HttpStatusCodes[HttpStatusCodes["NonAuthoritativeInformation"] = 203] = "NonAuthoritativeInformation";
    HttpStatusCodes[HttpStatusCodes["NoContent"] = 204] = "NoContent";
    HttpStatusCodes[HttpStatusCodes["ResetContent"] = 205] = "ResetContent";
    HttpStatusCodes[HttpStatusCodes["PartialContent"] = 206] = "PartialContent";
    HttpStatusCodes[HttpStatusCodes["MultiStatus"] = 207] = "MultiStatus";
    HttpStatusCodes[HttpStatusCodes["AlreadyReported"] = 208] = "AlreadyReported";
    HttpStatusCodes[HttpStatusCodes["IMUsed"] = 226] = "IMUsed";
    HttpStatusCodes[HttpStatusCodes["MultipleChoices"] = 300] = "MultipleChoices";
    HttpStatusCodes[HttpStatusCodes["MovedPermanently"] = 301] = "MovedPermanently";
    HttpStatusCodes[HttpStatusCodes["Found"] = 302] = "Found";
    HttpStatusCodes[HttpStatusCodes["SeeOther"] = 303] = "SeeOther";
    HttpStatusCodes[HttpStatusCodes["NotModified"] = 304] = "NotModified";
    HttpStatusCodes[HttpStatusCodes["UseProxy"] = 305] = "UseProxy";
    HttpStatusCodes[HttpStatusCodes["SwitchProxy"] = 306] = "SwitchProxy";
    HttpStatusCodes[HttpStatusCodes["TemporaryRedirect"] = 307] = "TemporaryRedirect";
    HttpStatusCodes[HttpStatusCodes["PermanentRedirect"] = 308] = "PermanentRedirect";
    HttpStatusCodes[HttpStatusCodes["BadRequest"] = 400] = "BadRequest";
    HttpStatusCodes[HttpStatusCodes["Unauthorized"] = 401] = "Unauthorized";
    HttpStatusCodes[HttpStatusCodes["PaymentRequired"] = 402] = "PaymentRequired";
    HttpStatusCodes[HttpStatusCodes["Forbidden"] = 403] = "Forbidden";
    HttpStatusCodes[HttpStatusCodes["NotFound"] = 404] = "NotFound";
    HttpStatusCodes[HttpStatusCodes["MethodNotAllowed"] = 405] = "MethodNotAllowed";
    HttpStatusCodes[HttpStatusCodes["NotAcceptable"] = 406] = "NotAcceptable";
    HttpStatusCodes[HttpStatusCodes["ProxyAuthenticationRequired"] = 407] = "ProxyAuthenticationRequired";
    HttpStatusCodes[HttpStatusCodes["RequestTimeout"] = 408] = "RequestTimeout";
    HttpStatusCodes[HttpStatusCodes["Conflict"] = 409] = "Conflict";
    HttpStatusCodes[HttpStatusCodes["Gone"] = 410] = "Gone";
    HttpStatusCodes[HttpStatusCodes["LengthRequired"] = 411] = "LengthRequired";
    HttpStatusCodes[HttpStatusCodes["PreconditionFailed"] = 412] = "PreconditionFailed";
    HttpStatusCodes[HttpStatusCodes["PayloadTooLarge"] = 413] = "PayloadTooLarge";
    HttpStatusCodes[HttpStatusCodes["URITooLong"] = 414] = "URITooLong";
    HttpStatusCodes[HttpStatusCodes["UnsupportedMediaType"] = 415] = "UnsupportedMediaType";
    HttpStatusCodes[HttpStatusCodes["RangeNotSatisfiable"] = 416] = "RangeNotSatisfiable";
    HttpStatusCodes[HttpStatusCodes["ExpectationFailed"] = 417] = "ExpectationFailed";
    HttpStatusCodes[HttpStatusCodes["ImATeapot"] = 418] = "ImATeapot";
    HttpStatusCodes[HttpStatusCodes["MethodFailure"] = 420] = "MethodFailure";
    HttpStatusCodes[HttpStatusCodes["EnhanceYourCalm"] = 420] = "EnhanceYourCalm";
    HttpStatusCodes[HttpStatusCodes["MisdirectedRequest"] = 421] = "MisdirectedRequest";
    HttpStatusCodes[HttpStatusCodes["UnprocessableEntity"] = 422] = "UnprocessableEntity";
    HttpStatusCodes[HttpStatusCodes["Locked"] = 423] = "Locked";
    HttpStatusCodes[HttpStatusCodes["FailedDependency"] = 424] = "FailedDependency";
    HttpStatusCodes[HttpStatusCodes["UpgradeRequired"] = 426] = "UpgradeRequired";
    HttpStatusCodes[HttpStatusCodes["PreconditionRequired"] = 428] = "PreconditionRequired";
    HttpStatusCodes[HttpStatusCodes["TooManyRequests"] = 429] = "TooManyRequests";
    HttpStatusCodes[HttpStatusCodes["RequestHeaderFieldsTooLarge"] = 431] = "RequestHeaderFieldsTooLarge";
    HttpStatusCodes[HttpStatusCodes["LoginTimeout"] = 440] = "LoginTimeout";
    HttpStatusCodes[HttpStatusCodes["NoResponse"] = 444] = "NoResponse";
    HttpStatusCodes[HttpStatusCodes["RetryWith"] = 449] = "RetryWith";
    HttpStatusCodes[HttpStatusCodes["BlockedByWindowsParentalControls"] = 450] = "BlockedByWindowsParentalControls";
    HttpStatusCodes[HttpStatusCodes["UnavailableForLegalReasons"] = 451] = "UnavailableForLegalReasons";
    HttpStatusCodes[HttpStatusCodes["Redirect"] = 451] = "Redirect";
    HttpStatusCodes[HttpStatusCodes["SSLCertificateError"] = 495] = "SSLCertificateError";
    HttpStatusCodes[HttpStatusCodes["SSLCertificateRequired"] = 496] = "SSLCertificateRequired";
    HttpStatusCodes[HttpStatusCodes["HttpRequestSentToHttpsPort"] = 497] = "HttpRequestSentToHttpsPort";
    HttpStatusCodes[HttpStatusCodes["ClientClosedRequest"] = 499] = "ClientClosedRequest";
    HttpStatusCodes[HttpStatusCodes["InvalidToken"] = 498] = "InvalidToken";
    HttpStatusCodes[HttpStatusCodes["TokenRequired"] = 499] = "TokenRequired";
    HttpStatusCodes[HttpStatusCodes["RequestWasForbiddenByAntivirus"] = 499] = "RequestWasForbiddenByAntivirus";
    HttpStatusCodes[HttpStatusCodes["InternalServerError"] = 500] = "InternalServerError";
    HttpStatusCodes[HttpStatusCodes["NotImplemented"] = 501] = "NotImplemented";
    HttpStatusCodes[HttpStatusCodes["BadGateway"] = 502] = "BadGateway";
    HttpStatusCodes[HttpStatusCodes["ServiceUnavailable"] = 503] = "ServiceUnavailable";
    HttpStatusCodes[HttpStatusCodes["GatewayTimeout"] = 504] = "GatewayTimeout";
    HttpStatusCodes[HttpStatusCodes["HttpVersionNotSupported"] = 505] = "HttpVersionNotSupported";
    HttpStatusCodes[HttpStatusCodes["VariantAlsoNegotiates"] = 506] = "VariantAlsoNegotiates";
    HttpStatusCodes[HttpStatusCodes["InsufficientStorage"] = 507] = "InsufficientStorage";
    HttpStatusCodes[HttpStatusCodes["LoopDetected"] = 508] = "LoopDetected";
    HttpStatusCodes[HttpStatusCodes["BandwidthLimitExceeded"] = 509] = "BandwidthLimitExceeded";
    HttpStatusCodes[HttpStatusCodes["NotExtended"] = 510] = "NotExtended";
    HttpStatusCodes[HttpStatusCodes["NetworkAuthenticationRequired"] = 511] = "NetworkAuthenticationRequired";
    HttpStatusCodes[HttpStatusCodes["UnknownError"] = 520] = "UnknownError";
    HttpStatusCodes[HttpStatusCodes["WebServerIsDown"] = 521] = "WebServerIsDown";
    HttpStatusCodes[HttpStatusCodes["ConnectionTimedOut"] = 522] = "ConnectionTimedOut";
    HttpStatusCodes[HttpStatusCodes["OriginIsUnreachable"] = 523] = "OriginIsUnreachable";
    HttpStatusCodes[HttpStatusCodes["ATimeoutOccurred"] = 524] = "ATimeoutOccurred";
    HttpStatusCodes[HttpStatusCodes["SSLHandshakeFailed"] = 525] = "SSLHandshakeFailed";
    HttpStatusCodes[HttpStatusCodes["InvalidSSLCertificate"] = 526] = "InvalidSSLCertificate";
    HttpStatusCodes[HttpStatusCodes["SiteIsFrozen"] = 530] = "SiteIsFrozen";
})(exports.HttpStatusCodes || (exports.HttpStatusCodes = {}));
var HttpStatusCodes = exports.HttpStatusCodes;
